src/github.com/mozilla-services/pushgo/simplepush/dynamo_store.go

```diff
@@ -0,0 +1,507 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package simplepush
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/jrconlin/goamz/aws"
+	"github.com/jrconlin/goamz/dynamodb"
+)
+
+const (
+	DB_LABEL   = "dynamodb"
+	UAID_LABEL = "uaid"
+	CHID_LABEL = "chid"
+	PING_LABEL = "proprietary_ping"
+	VERS_LABEL = "version"
+	MODD_LABEL = "modified"
+)
+
+var (
+	ErrInvalidRegion   = errors.New("Invalid Region")
+	ErrDynamoDBTimeout = errors.New("DynamoDB function timed out")
+	ErrDynamoDBFailure = errors.New("DynamoDB returned non success")
+)
+
+type DynamoDBConf struct {
+	// Name of the table to use
+	TableName string `toml:"tablename" env:"tablename"`
+
+	// number of provisioned hosts (more hosts = more money)
+	ReadProv  int64 `toml:"read_cap_units" env:"read_cap_units"`
+	WriteProv int64 `toml:"write_cap_units" env:"write_cap_units"`
+
+	// Region that the dbstore is in
+	Region string `toml:"region" env:"region"`
+
+	// Auth info (using aws.GetAuth)
+	// NOTE: if using environment variables, please use goamz Var names.
+	// AWS_ACCESS_KEY
+	// AWS_SECRET_KEY
+	Access string `toml:"accesskey" `
+	Secret string `toml:"secret" `
+
+	MaxChannels int `toml:"max_channels" env:"max_channels"`
+}
+
+type dynamoTable interface {
+	DescribeTable() (*dynamodb.TableDescriptionT, error)
+	CountQuery([]dynamodb.AttributeComparison) (int64, error)
+	PutItem(string, string, []dynamodb.Attribute) (bool, error)
+	UpdateAttributes(*dynamodb.Key, []dynamodb.Attribute) (bool, error)
+	DeleteItem(*dynamodb.Key) (bool, error)
+	Query([]dynamodb.AttributeComparison) ([]map[string]*dynamodb.Attribute, error)
+	DeleteAttributes(*dynamodb.Key, []dynamodb.Attribute) (bool, error)
+}
+
+type dynamoServer interface {
+	NewTable(string, dynamodb.PrimaryKey) *dynamodb.Table
+	CreateTable(dynamodb.TableDescriptionT) (string, error)
+	ListTables() ([]string, error)
+}
+
+type DynamoDBStore struct {
+	logger        *SimpleLogger
+	region        aws.Region
+	auth          aws.Auth
+	pk            dynamodb.PrimaryKey
+	server        dynamoServer
+	table         dynamoTable
+	tablename     string
+	readProv      int64
+	writeProv     int64
+	closeSignal   chan bool
+	maxChannels   int
+	statusTimeout time.Duration
+	statusIdle    time.Duration
+}
+
+const (
+	RFC_ISO_8601  = "20060102T150405Z"
+	DYNAMO_PREFIX = "DynamoDB_20120810"
+)
+
+func (s *DynamoDBStore) waitUntilStatus(table dynamoTable, status string) (err error) {
+	done := make(chan bool)
+	timeout := time.After(s.statusTimeout)
+	go func() {
+		for {
```

Comment:
Oh, I see...`done` is unbuffered, so the send on line 105 will block until it's matched by a receive on line 113. Otherwise, if the receive times out, `done` will be closed, signaling the loop to exit on the next turn. That took a few tries to parse correctly. :grin: